### PR TITLE
KAFKA-16161: Avoid empty remote metadata snapshot file in partition dir

### PR DESCRIPTION
**Pull request by kamalcph**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by showuon — COMMENTED** [review]
@kamalcph , thanks for the PR. I am thinking should we remove `FileBasedRemoteLogMetadataCache` and `RemoteLogMetadataSnapshotFile`, and use `RemoteLogMetadataCache` directly, because both of them are unneeded? We can create (or already existing?) a JIRA ticket to record this PR for future reference (to add them back in the future). WDYT?

**Comment by kamalcph** [comment]
Left the [RemoteLogSegmentMetadataSnapshotRecord.json](https://sourcegraph.com/github.com/apache/kafka@40e87ae35beb389d6419d32130174d7c68fa4d19/-/blob/storage/src/main/resources/message/RemoteLogSegmentMetadataRecord.json) and it's related changes.

**Review by showuon — APPROVED** [review]
LGTM! 1 question question: Did we have JIRA ticket for re-enabling the snapshot? If no, should we create one?

**Comment by kamalcph** [comment]
> LGTM! 1 question question: Did we have JIRA ticket for re-enabling the snapshot? If no, should we create one?

Filed https://issues.apache.org/jira/browse/KAFKA-16454 to track this task.

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCache.java [hunk 1]
```diff
@@ -115,6 +115,9 @@ public boolean isInitialized() {
         return initializedLatch.getCount() == 0;
     }
 
+    public RemoteLogMetadataCache() {
```
Comment (chia7712) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCache.java]: Is this necessary? the constructor without parameters is the default constructor to this class.

**Review by chia7712 — APPROVED** [review]
LGTM

maybe we should update comments on https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogSegmentMetadataSnapshot.java#L39

**Review by chia7712 — APPROVED** [review]

**Review by satishd — APPROVED** [review]
Thanks @kamalcph for the PR. LGTM.

**Comment by showuon** [comment]
Failed tests are unrelated.

**Pull request merged**